csds-client/client/client.go

```diff
@@ -0,0 +1,255 @@
+		}
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.info.jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return nil
+		}
+	case "auto":
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.info.uri {
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := getValueByKeyFromNodeMatcher(c.nm, key); projectNum != "" {
+				c.md = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("Auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return fmt.Errorf("Invalid authn_mode")
+	}
+}
+
+// New creates a new client
+func New() (*Client, error) {
+	c := &Client{
+		info: ParseFlags(),
+	}
+	if c.info.platform != "gcp" {
+		return c, fmt.Errorf("Can not support this platform now")
+	}
+	if c.info.apiVersion != "v2" {
+		return c, fmt.Errorf("Can not suppoort this api version now")
+	}
+
+	if err := c.parseNodeMatcher(); err != nil {
+		return c, err
+	}
+
+	return c, nil
+}
+
+// Run connects the client to the uri and call doRequest based on monitor_freq flag
+func (c *Client) Run() error {
+	if err := c.connWithAuth(); err != nil {
+		return err
+	}
+	defer c.cc.Close()
+
+	c.csdsClient = csdspb.NewClientStatusDiscoveryServiceClient(c.cc)
+	var ctx context.Context
+	if c.md != nil {
+		ctx = metadata.NewOutgoingContext(context.Background(), c.md)
+	} else {
+		ctx = context.Background()
+	}
+	streamClientStatus, err := c.csdsClient.StreamClientStatus(ctx)
+	if err != nil {
+		return err
+	}
+
+	// run once or run with monitor mode
+	switch c.info.monitorFreq {
+	case "":
+		if err := c.doRequest(streamClientStatus); err != nil {
+			return err
+		}
+		return nil
+	default:
+		freq, err := time.ParseDuration(c.info.monitorFreq)
+		if err != nil {
+			return err
+		}
+		ticker := time.NewTicker(freq)
+
+		// keep track of 'ctrl+c' to stop
+		s := make(chan os.Signal)
+		signal.Notify(s, os.Interrupt, syscall.SIGTERM)
+		go func() {
```

Comment:
Why do you need a go routine here?

Reply:
Just want to keep track of `Ctrl-C` and then print out `Client Stopped` message rather than just interrupting the client. If this is not very useful, I will remove it.

Reply:
probably not needed.

Reply:
Removed.